### PR TITLE
chore: use coreutils hashsum

### PR DESCRIPTION
**Pull request by thesayyn**

### CODE REVIEW

**Comment by gregmagolan** [comment]
Will also need to add a test for hashes to e2e/bzlmod with this one :)

**Comment by thesayyn** [comment]
adding derek as a reviewer as he knows the repo and bzlmod

**Comment by kormide** [comment]
I think you could add to `e2e/workspace` and `e2e/bzlmod`, just one usage (e.g., a genrule) as a smoke test that workspace/e2e users are able to use the toolchain. For the most part unit tests can exercise coreutils, and eventually we'll be running all unit tests with bzlmod enabled on ci, so you don't need to worry about exercising it fully in an e2e.

File: e2e/bzlmod_coreutils/BUILD.bazel [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+load("@aspect_bazel_lib//tools:hashes.bzl", "hashes")
+
+genrule(
+    name = "ls",
+    outs = ["ls.txt"],
+    cmd = "$(COREUTILS_BIN) ls > $@",
+    toolchains = ["@coreutils_toolchains//:resolved_toolchain"],
+)
+
+hashes(
```
Comment (gregmagolan) [inline e2e/bzlmod_coreutils/BUILD.bazel:11]: lets just move these to e2e/bzlmod which is already testing out the other toolchains

File: tools/hashes.bzl [hunk 2]
```diff
@@ -6,36 +6,32 @@ via output groups.
...
+    out = ctx.actions.declare_file("{}.{}".format(file.path, algo))
     ctx.actions.run_shell(
-        outputs = [md5out],
-        inputs = [ctx.file.src],
-        command = "ROOT=$PWD && cd {} && md5sum {} > $ROOT/{}".format(ctx.file.src.dirname, ctx.file.src.basename, md5out.path),
+        outputs = [out],
+        inputs = [file],
+        tools = [coreutils.coreutils_info.bin],
+        # coreutils has --no-names option but it doesn't work in current version, so we have to use cut.
+        command = """HASH=$({coreutils} hashsum --{algo} {src} | {coreutils} cut -f1 -d " ") && {coreutils} echo -ne "$HASH {basename}" > {out}""".format(
```
Comment (gregmagolan) [inline tools/hashes.bzl:18]: 🧙

**Comment by gregmagolan** [comment]
> 

Yup. Agreed that e2e/bzlmod & e2e/workspace are just smoke tests. Since we're not in a rush to land this @thesayyn we may as well take the opportunity to combine e2e/workspace & e2e/bzlmod into a single e2e/workspace and add the `bzlmodEnabled: [true, false]` pattern to CI for bazel-lib which @kormide just rolled out in rules_js: https://github.com/aspect-build/rules_js/blob/047decb02a7e1497d23cfb9b239b56778ea19f4d/.github/workflows/ci.yaml#L132

**Comment by gregmagolan** [comment]
@alexeagle is going to configure our preferred pattern for bzlmod/workspace smoketest e2e in rules_swc. Lets wait until that lands and then copy it over here.

**Comment by alexeagle** [comment]
Working on that right now...

**Comment by alexeagle** [comment]
https://github.com/aspect-build/rules_swc/pull/153

**Comment by alexeagle** [comment]
@thesayyn should be unblocked now

**Comment by thesayyn** [comment]
oh, I missed this one. I'll takeover thanks!

**Comment by thesayyn** [comment]
now, this is more of a coreutils + e2e cleanup.

**Comment by thesayyn** [comment]
this is ready now

**Comment by kormide** [comment]
* Can we rename `e2e/workspace` to `e2e/smoke`? (`.bcr/presubmit.yml` also needs to reference this target)`.
* Each e2e should also have a `WORKSPACE.bzlmod` file, which tells Bazel to ignore `WORKSPACE` when bzlmod is enabled, and any repository rules can't be set up under MODULE.bazel go in there.

File: e2e/workspace/MODULE.bazel [hunk 3]
```diff
@@ -0,0 +1,14 @@
+module(
+    name = "workspace",
```
Comment (kormide) [inline e2e/workspace/MODULE.bazel:2]: Can we rename this to `e2e/smoke`, include an equivalent WORKSPACE file, and update `.bcr/presubmit` to point to it?

Comment (thesayyn) [reply e2e/workspace/MODULE.bazel:2]: done

File: e2e/copy_to_directory/MODULE.bazel [hunk 4]
```diff
@@ -1,5 +1,5 @@
 module(
-    name = "e2e_bzlmod",
+    name = "workspace",
```
Comment (kormide) [inline e2e/copy_to_directory/MODULE.bazel:2]: Note that you can leave out the module name for e2es.

**Review by kormide — APPROVED** [review]

File: e2e/copy_to_directory/MODULE.bazel (hunk 4 above)
Comment (thesayyn) [reply e2e/copy_to_directory/MODULE.bazel]: fixed these as well.

**Comment by alexeagle** [comment]
This broke the windows build on `main`ppp

**Comment by gregmagolan** [comment]
Looks like a small typo. https://github.com/aspect-build/bazel-lib/pull/395 to fix

**Pull request merged**